libibverbs/device.c

```diff
@@ -262,6 +267,8 @@ struct ibv_context *__ibv_open_device(struct ibv_device *device)
 	context->cmd_fd = cmd_fd;
 	pthread_mutex_init(&context->mutex, NULL);
 
+	ibverbs_device_hold(device);
+
```

Comment:
> Why do we need a solution for everything that touches sysfs ? same TS on the same sysfs device name entry should guaranteed same device .

It is the same basic problem, hotplug could re-use the sysfs files between charddev open and sysfs read, so that needs to be protected by a TS check also. This is why openat might be nice, instead of passing uverbs_sys_path as a char *, pass in a fd for  the base sysfs dir and use openat.  The common code can confirm that the dir fd was opened without races, and the kernel will not allow a new driver to 'take over' an open dir fd, AFAIK. So if there is a race then openat will just fail not return bogus data.

> In addition, extra checking before attempting to open looks redundant, as just after checking 

It seems necessary to get a consistent and proper errno code in all cases.